File: crates/lucidos-app/src/store/actions/form-requests.ts
import { activeInlineForm, panelOverlay, showToast } from '../store';
import type { FormRequestEvent, FormRequestOutcome } from '../thread-events/thread-event-types';
import type { MessageOrigin } from '../../generated/thread-event-wire';
import { listPendingFormRequests } from '../../api/client/settings';
import { errorDetail } from '../../utils/errorDetail';
import { openCredentialRequest } from './credentials';
import { openPluginInstallRequest } from './plugin-install';
import { openPluginUninstallRequest } from './plugin-uninstall';
import { openEmailConfirmRequest } from './email-confirm';
import { getDeviceId } from './devices';
import {
  handleNavigationRequest,
  navigationIsForThisDevice,
  routeThreadNavigation,
} from './navigation-request';
import { formatThreadLabel } from './thread-label';

/** *Form requests*: something the agent asked the user to act on, persisted
 *  until a `FormRequestResolved` closes it (engine `engine/form_requests.rs`).
 *
 *  A live frame opens its form at once. This page can miss a frame: a replaced
 *  stream, a reconnect gap, a lagged broadcast, a reload. So
 *  `syncPendingFormRequests` runs on every stream open to find what it missed.
 *  The transcript row reopens one on demand. */

/** Requests this page load has already put in front of the user. A sync offers
 *  only what is not here, so a refocus never re-pops a form the user walked
 *  away from. A reload starts empty, and offers the open ones again. */
const offered = new Set<string>();

/** Test seam: forget what this page has offered. */
export function resetOfferedFormRequests(): void {
  offered.clear();
}

/** Open the form `event` asks for. The live frame, the pending sync and the
 *  transcript's Open button all come here.
 *
 *  `byUser` is the Open button. An authorization page the engine sent goes to
 *  one device, and never hijacks another thread's page. A user who pressed Open
 *  asked for it here and now. */
export function openFormRequest(
  threadId: string,
  event: FormRequestEvent,
  { byUser = false }: { byUser?: boolean } = {},
): void {
  offered.add(event.request_id);
  let payload: Record<string, unknown>;
  try {
    payload = JSON.parse(event.payload) as Record<string, unknown>;
  } catch (e) {
    showToast(`Failed to open the ${describeFormRequest(event)}: ${errorDetail(e)}`, 'error');
    return;
  }
  switch (event.type) {
    case 'CredentialRequested':
      openCredentialRequest({ ...payload, form_request_id: event.request_id });
      return;
    case 'EmailConfirmRequested':
      openEmailConfirmRequest({
        ...(payload as unknown as Parameters<typeof openEmailConfirmRequest>[0]),
        form_request_id: event.request_id,
      });
      return;
    case 'PluginInstallRequested':
      openPluginInstallRequest(payload as unknown as Parameters<typeof openPluginInstallRequest>[0]);
      return;
    case 'PluginUninstallRequested':
      openPluginUninstallRequest(
        payload as unknown as Parameters<typeof openPluginUninstallRequest>[0],
      );
      return;
    case 'OAuthAuthorizationRequested': {
      const nav = payload as unknown as Parameters<typeof handleNavigationRequest>[0];
      if (byUser) handleNavigationRequest(nav, { source: formatThreadLabel(threadId) });
      else routeThreadNavigation(nav, event.actor, threadId);
      return;
    }
  }
}

/** Whether this device should open `event` unasked. An authorization page goes
 *  to the device its actor names, as a navigation does. A form goes to every
 *  device, and the first answer closes it on the others. */
function isForThisDevice(event: FormRequestEvent): boolean {
  if (event.type !== 'OAuthAuthorizationRequested') return true;
  return navigationIsForThisDevice(event.actor);
}

/** Read the open requests and offer the newest one this page has not offered.
 *
 *  Called on every stream open, after the open, so a request emitted between
 *  the read and the open still arrives live. A form the user has open is
 *  never replaced: the rest stay reachable from their transcript rows. */
export async function syncPendingFormRequests(): Promise<void> {
  let pending;
  try {
    pending = await listPendingFormRequests();
  } catch (e) {
    // Runs on every stream open with no user action behind it. The next open
    // retries, and each open request stays reachable from its transcript row.
    // A toast here would only repeat itself on a flaky connection.
    console.warn('[FormRequests] reading the open form requests failed:', e);
    return;
  }
  if (activeInlineForm.value) return;
  const fresh = pending.filter((p) => !offered.has(p.request_id) && isForThisDevice(p.event));
  const newest = fresh[fresh.length - 1];
  if (newest) openFormRequest(newest.thread_id, newest.event);
}

/** Close the open form a resolved request was showing, when another device or
 *  path answered it.
 *
 *  **The acting device is exempt by its ACTOR.** Its own save, send or confirm
 *  already handles the form. A receipt it is about to stamp must not lose the
 *  race to this frame, and a receipt already on screen is left alone too.
 *
 *  A supersede is no such answer. A message typed on this device replaces the
 *  form, and nothing else here would close it. */
export function closeResolvedFormRequest(
  requestId: string,
  outcome: FormRequestOutcome,
  actor: MessageOrigin | null | undefined,
): void {
  const answeredHere = actor?.kind === 'device' && actor.device_id === getDeviceId();
  if (answeredHere && outcome !== 'superseded') return;
  const form = activeInlineForm.value;
  if (!form) return;
  const showing =
    (form.type === 'credential' && form.request?.form_request_id === requestId)
    || (form.type === 'email-confirm' && !form.sentAt && form.request.form_request_id === requestId)
    || (form.type === 'plugin-install' && !form.installed && form.request.install_id === requestId)
    || (form.type === 'plugin-uninstall' && !form.removed && form.request.uninstall_id === requestId);
  if (showing) panelOverlay.value = null;
}

/** The request in words, for a transcript row and an error toast. */
export function describeFormRequest(event: FormRequestEvent): string {
  switch (event.type) {
    case 'CredentialRequested': {
      const service = payloadField(event, 'service');
      return service ? `credential form for ${service}` : 'credential form';
    }
    case 'EmailConfirmRequested':
      return 'email confirmation';
    case 'PluginInstallRequested': {
      const name = payloadField(event, 'plugin_name') ?? payloadField(event, 'plugin_id');
      return name ? `install panel for ${name}` : 'plugin install panel';
    }
    case 'PluginUninstallRequested': {
      const name = payloadField(event, 'plugin_name') ?? payloadField(event, 'plugin_id');
      return name ? `uninstall panel for ${name}` : 'plugin uninstall panel';
    }
    case 'OAuthAuthorizationRequested':
      return 'authorization page';
  }
}

function payloadField(event: FormRequestEvent, key: string): string | undefined {
  try {
    const value = (JSON.parse(event.payload) as Record<string, unknown>)[key];
    return typeof value === 'string' && value ? value : undefined;
  } catch {
    // A payload that does not parse still has a generic name. Opening it is
    // where the parse failure is reported, with a toast.
    return undefined;
  }
}

File: crates/lucidos-engine/src/engine/form_requests.rs
//! *Form requests*: a request from the agent that the user must act on.
//!
//! Five `ThreadEvent` variants open one (`FORM_REQUEST_EVENT_TYPES`), and one
//! `FormRequestResolved` closes it, keyed on `request_id`. Both are persisted,
//! so an unanswered request survives a lost stream frame, a reload and an
//! engine restart. The client reads the open ones from
//! `GET /api/v1/form-requests/pending` on every stream open.
//!
//! This module is the one place that decides whether a request is still open.
//! See `docs/plans/2026-09-24-form-requests-survive-a-reconnect.md`.

use serde::Serialize;
use sqlx::PgPool;
use uuid::Uuid;

use crate::engine::event_bus::{BusEvent, EventBus};
use crate::engine::thread_events::{EventMeta, FormRequestOutcome, MessageOrigin, ThreadEvent};

type BoxError = Box<dyn std::error::Error + Send + Sync>;

/// The request variants. Each carries `request_id` and a JSON `payload`.
pub const FORM_REQUEST_EVENT_TYPES: &[&str] = &[
    "CredentialRequested",
    "PluginInstallRequested",
    "PluginUninstallRequested",
    "EmailConfirmRequested",
    "OAuthAuthorizationRequested",
];

/// The kinds whose open state lives in engine memory, and so dies with it:
/// plugin staging and the OAuth callback listener. Expired at boot.
const MEMORY_BACKED_EVENT_TYPES: &[&str] = &[
    "PluginInstallRequested",
    "PluginUninstallRequested",
    "OAuthAuthorizationRequested",
];

/// Serializes every resolve, so a check and its emit cannot interleave with
/// another resolve of the same request. One engine owns a workspace database.
static RESOLVE_LOCK: std::sync::LazyLock<tokio::sync::Mutex<()>> =
    std::sync::LazyLock::new(|| tokio::sync::Mutex::new(()));

/// An open form request, as `GET /api/v1/form-requests/pending` serves it.
#[derive(Debug, Clone, Serialize)]
pub struct PendingFormRequest {
    pub thread_id: Uuid,
    pub request_id: Uuid,
    /// The request event as the stream carries it: `type` plus its payload,
    /// meta fields (`actor`, `request_event_id`) included.
    pub event: serde_json::Value,
}

/// SQL that is true when the request row `e` has a resolution.
const RESOLVED_SQL: &str = "EXISTS ( \
     SELECT 1 FROM events r \
     WHERE r.event_type = 'FormRequestResolved' \
       AND r.payload->>'request_id' = e.payload->>'request_id')";

/// Every open form request in a thread the user can still see: not archived,
/// not discarded. Oldest first.
pub async fn pending(pool: &PgPool) -> Result<Vec<PendingFormRequest>, sqlx::Error> {
    let rows: Vec<(Uuid, Option<String>, serde_json::Value)> = sqlx::query_as(&format!(
        "SELECT e.thread_id, e.payload->>'request_id', \
                e.payload || jsonb_build_object('type', e.event_type) \
         FROM events e \
         JOIN thread_summaries ts ON ts.thread_id = e.thread_id \
         WHERE e.event_type = ANY($1) \
           AND ts.archive_state <> 'archived' \
           AND NOT {RESOLVED_SQL} \
         ORDER BY e.created, e.sequence"
    ))
    .bind(FORM_REQUEST_EVENT_TYPES)
    .fetch_all(pool)
    .await?;
    Ok(rows
        .into_iter()
        .filter_map(|(thread_id, request_id, event)| {
            let request_id = request_id.as_deref().and_then(|s| s.parse().ok())?;
            Some(PendingFormRequest {
                thread_id,
                request_id,
                event,
            })
        })
        .collect())
}

/// One open request row, for the in-engine sweeps.
struct OpenRequest {
    request_id: Uuid,
    event_type: String,
    /// The request's own `payload` string, parsed.
    body: serde_json::Value,
}

/// Open requests of the given types, in one thread or in all of them.
async fn open_requests(
    pool: &PgPool,
    thread_id: Option<Uuid>,
    types: &[&str],
) -> Result<Vec<OpenRequest>, sqlx::Error> {
    let rows: Vec<(Option<String>, String, Option<String>)> = sqlx::query_as(&format!(
        "SELECT e.payload->>'request_id', e.event_type, e.payload->>'payload' \
         FROM events e \
         WHERE e.event_type = ANY($1) \
           AND ($2::uuid IS NULL OR e.thread_id = $2) \
           AND NOT {RESOLVED_SQL}"
    ))
    .bind(types)
    .bind(thread_id)
    .fetch_all(pool)
    .await?;
    Ok(rows
        .into_iter()
        .filter_map(|(request_id, event_type, payload)| {
            Some(OpenRequest {
                request_id: request_id.as_deref()?.parse().ok()?,
                event_type,
                body: payload
                    .as_deref()
                    .and_then(|p| serde_json::from_str(p).ok())
                    .unwrap_or(serde_json::Value::Null),
            })
        })
        .collect())
}

/// The variant name of the form request `request_id` names, if any does.
pub async fn event_type_of(pool: &PgPool, request_id: Uuid) -> Result<Option<String>, sqlx::Error> {
    sqlx::query_scalar(
        "SELECT event_type FROM events \
         WHERE event_type = ANY($1) AND payload->>'request_id' = $2 \
         LIMIT 1",
    )
    .bind(FORM_REQUEST_EVENT_TYPES)
    .bind(request_id.to_string())
    .fetch_optional(pool)
    .await
}

/// Close `request_id` with `outcome`, once.
///
/// Returns whether this call emitted the resolution. `false` means there is
/// nothing to close: no form request has that id (a plugin staged over HTTP, a
/// Settings-initiated OAuth flow), or it is already resolved.
pub async fn resolve(
    pool: &PgPool,
    bus: &EventBus,
    request_id: Uuid,
    outcome: FormRequestOutcome,
    actor: Option<MessageOrigin>,
) -> Result<bool, BoxError> {
    let _serialized = RESOLVE_LOCK.lock().await;
    let row: Option<(Uuid, bool)> = sqlx::query_as(&format!(
        "SELECT e.thread_id, {RESOLVED_SQL} \
         FROM events e \
         WHERE e.event_type = ANY($1) \
           AND e.payload->>'request_id' = $2 \
         LIMIT 1"
    ))
    .bind(FORM_REQUEST_EVENT_TYPES)
    .bind(request_id.to_string())
    .fetch_optional(pool)
    .await?;
    let Some((thread_id, false)) = row else {
        return Ok(false);
    };
    bus.emit(BusEvent::Thread {
        thread_id,
        event: ThreadEvent::FormRequestResolved {
            request_id,
            outcome,
        },
        meta: EventMeta::with_actor(actor),
    })
    .await?;
    Ok(true)
}

/// [`resolve`] for a caller that has no one to report a failure to: a sweep, a
/// listener task, an emit site whose own work already succeeded.
pub async fn resolve_or_log(
    pool: &PgPool,
    bus: &EventBus,
    request_id: Uuid,
    outcome: FormRequestOutcome,
    actor: Option<MessageOrigin>,
) {
    if let Err(e) = resolve(pool, bus, request_id, outcome, actor).await {
        crate::log!("[FormRequests] resolving {request_id} as {outcome:?} failed: {e}");
    }
}

/// Complete `request_id` when it names a request of `kind`. For a save or a
/// send that says which form it answers. The work itself already succeeded, so
/// a failure here is logged rather than reported.
pub async fn complete_named_request(
    pool: &PgPool,
    bus: &EventBus,
    request_id: Uuid,
    kind: &str,
    actor: Option<MessageOrigin>,
) {
    match event_type_of(pool, request_id).await {
        Ok(Some(found)) if found == kind => {
            resolve_or_log(pool, bus, request_id, FormRequestOutcome::Completed, actor).await
        }
        Ok(found) => {
            crate::log!("[FormRequests] {request_id} is {found:?}, not a {kind}; left open")
        }
        Err(e) => crate::log!("[FormRequests] looking up {request_id} failed: {e}"),
    }
}

/// Resolve every open request `pick` selects. For the sweeps.
async fn resolve_open_where(
    pool: &PgPool,
    bus: &EventBus,
    thread_id: Option<Uuid>,
    types: &[&str],
    outcome: FormRequestOutcome,
    actor: Option<MessageOrigin>,
    pick: impl Fn(&OpenRequest) -> bool,
) {
    let open = match open_requests(pool, thread_id, types).await {
        Ok(open) => open,
        Err(e) => {
            crate::log!("[FormRequests] open-request query failed: {e}");
            return;
        }
    };
    for request in open.iter().filter(|r| pick(r)) {
        resolve_or_log(pool, bus, request.request_id, outcome, actor.clone()).await;
    }
}

/// What a request is about, when a newer one for the same thing replaces it.
/// `None` for a kind where two open at once are both valid (two emails).
fn subject(event_type: &str, body: &serde_json::Value) -> Option<String> {
    let key = match event_type {
        "CredentialRequested" => "service",
        "PluginInstallRequested" | "PluginUninstallRequested" => "plugin_id",
        _ => return None,
    };
    body[key].as_str().map(str::to_string)
}

/// Supersede the thread's open requests for the same subject as `event`.
/// Run just before `event` is emitted, so only the newest stays open.
pub async fn supersede_same_subject(
    pool: &PgPool,
    bus: &EventBus,
    thread_id: Uuid,
    event: &ThreadEvent,
) {
    let (event_type, payload) = match event {
        ThreadEvent::CredentialRequested { payload, .. }
        | ThreadEvent::PluginInstallRequested { payload, .. }
        | ThreadEvent::PluginUninstallRequested { payload, .. } => (event.event_type(), payload),
        _ => return,
    };
    let body = serde_json::from_str(payload).unwrap_or(serde_json::Value::Null);
    let Some(wanted) = subject(event_type, &body) else {
        return;
    };
    resolve_open_where(
        pool,
        bus,
        Some(thread_id),
        &[event_type],
        FormRequestOutcome::Superseded,
        None,
        |open| subject(&open.event_type, &open.body).as_deref() == Some(wanted.as_str()),
    )
    .await;
}

/// Emit a new form request, superseding the older ones it replaces first.
/// The one emitter for a sentinel-born request, shared by both agentic loops.
pub async fn emit_request(
    pool: &PgPool,
    bus: &EventBus,
    thread_id: Uuid,
    event: ThreadEvent,
    meta: EventMeta,
    label: &str,
) {
    supersede_same_subject(pool, bus, thread_id, &event).await;
    bus.emit_or_log(
        BusEvent::Thread {
            thread_id,
            event,
            meta,
        },
        label,
    )
    .await;
}

/// A new user message answers the thread's open forms, as it does a permission
/// card. The OAuth page is left to its own flow, which is still listening.
pub async fn supersede_on_user_message(
    pool: &PgPool,
    bus: &EventBus,
    thread_id: Uuid,
    actor: Option<MessageOrigin>,
) {
    resolve_open_where(
        pool,
        bus,
        Some(thread_id),
        &[
            "CredentialRequested",
            "PluginInstallRequested",
            "PluginUninstallRequested",
            "EmailConfirmRequested",
        ],
        FormRequestOutcome::Superseded,
        actor,
        |_| true,
    )
    .await;
}

/// A saved credential answers every open request for its service, whichever
/// form or device saved it.
pub async fn complete_credential_requests_for(
    pool: &PgPool,
    bus: &EventBus,
    service_name: &str,
    actor: Option<MessageOrigin>,
) {
    resolve_open_where(
        pool,
        bus,
        None,
        &["CredentialRequested"],
        FormRequestOutcome::Completed,
        actor,
        |open| open.body["service"].as_str() == Some(service_name),
    )
    .await;
}

/// Whether `request` can still be answered, as far as engine memory can say.
///
/// A plugin request whose staging is gone cannot: its Confirm would 404. It is
/// filtered out rather than resolved, because a Confirm in flight pops the
/// staging before it records its own outcome.
pub fn is_answerable(request: &PendingFormRequest, is_staged: impl Fn(&str, Uuid) -> bool) -> bool {
    match request.event["type"].as_str() {
        Some(kind @ ("PluginInstallRequested" | "PluginUninstallRequested")) => {
            is_staged(kind, request.request_id)
        }
        _ => true,
    }
}

/// Boot: expire every open request whose state lived in engine memory. The
/// staging and the listeners died with the previous engine.
pub async fn expire_memory_backed_requests(pool: &PgPool, bus: &EventBus) {
    resolve_open_where(
        pool,
        bus,
        None,
        MEMORY_BACKED_EVENT_TYPES,
        FormRequestOutcome::Expired,
        None,
        |_| true,
    )
    .await;
}

#[cfg(test)]
#[path = "form_requests_tests.rs"]
mod tests;
